pkg/common/consumer/managed_group.go

```diff
@@ -18,50 +18,151 @@ package consumer
 
 import (
 	"context"
+	"fmt"
+	"sync"
 	"sync/atomic"
 	"time"
 
+	"knative.dev/eventing-kafka/pkg/common/controlprotocol/commands"
+
 	"github.com/Shopify/sarama"
 	"go.uber.org/zap"
 )
 
+// GroupLockedError is the error returned if the a locked managed group is given a different token for access
+var GroupLockedError = fmt.Errorf("managed group lock failed: locked by a different token")
+
 // managedGroup contains information about a Sarama ConsumerGroup that is required to start (i.e. re-create) it
-type managedGroup struct {
+type managedGroup interface {
+	consume(context.Context, []string, sarama.ConsumerGroupHandler) error
+	start(sarama.ConsumerGroup)
+	stop() error
+	close() error
+	errors() chan error
+	processLock(*commands.CommandLock, bool) error
+}
+
+// managedGroupImpl implements the managedGroup interface
+type managedGroupImpl struct {
 	logger             *zap.Logger
 	group              sarama.ConsumerGroup // The Sarama ConsumerGroup which is under management
-	errors             chan error           // An error channel that will replicate the errors from the Sarama ConsumerGroup
+	groupErrors        chan error           // An error channel that will replicate the errors from the Sarama ConsumerGroup
 	restartWaitChannel chan struct{}        // A channel that will be closed when a stopped group is restarted
 	stopped            atomic.Value         // Boolean value indicating that the managed group is stopped
 	cancelErrors       func()               // Called by the manager's CloseConsumerGroup to terminate the error forwarding
 	cancelConsume      func()               // Called by the manager during CloseConsumerGroup
 	lockedBy           atomic.Value         // The LockToken of the ConsumerGroupAsyncCommand that requested the lock
 	cancelLockTimeout  func()               // Called internally to stop the lock timeout when a lock is removed
+	groupMutex         sync.Mutex           // Used to synchronize access to the internal sarama ConsumerGroup
 }
 
 // createManagedGroup associates a Sarama ConsumerGroup and cancel function (usually from the factory)
 // inside a new managedGroup struct.  If a timeout is given (nonzero), the lockId will be reset to an
 // empty string (i.e. "unlocked") after that time has passed.
-func createManagedGroup(logger *zap.Logger, group sarama.ConsumerGroup, cancel func()) *managedGroup {
+func createManagedGroup(ctx context.Context, logger *zap.Logger, group sarama.ConsumerGroup, cancelErrors func(), cancelConsume func()) managedGroup {
 
-	managedGrp := managedGroup{
-		logger:       logger,
-		group:        group,
-		errors:       make(chan error),
-		cancelErrors: cancel,
+	managedGrp := managedGroupImpl{
```

Comment:
not to nit pick but... it seems a bit more "standard / idiomatic" to do `managedGrp := &managedGroupImpl{}` here instead of returning the address below?

Reply:
I'll buy that